--- mac80211.sh		2011-04-30 16:51:32.739529466 +0200
+++ mac80211.sh_qmp	2011-05-04 17:21:10.879532488 +0200
@@ -418,6 +418,7 @@ enable_mac80211() {
 					config_get encryption "$vif" encryption
 					config_get key "$vif" key 1
 					config_get mcast_rate "$vif" mcast_rate
+					config_get htmode "$device" htmode
 
 					local keyspec=""
 					[ "$encryption" == "wep" ] && {
@@ -457,7 +458,7 @@ enable_mac80211() {
 						[ "$mcsub" -gt 0 ] && mcval="$mcval.$mcsub"
 					}
 
-					iw dev "$ifname" ibss join "$ssid" $freq \
+					iw dev "$ifname" ibss join "$ssid" $freq $htmode \
 						${fixed:+fixed-freq} $bssid \
 						${beacon_int:+beacon-interval $beacon_int} \
 						${brstr:+basic-rates $brstr} \
