--- bmx-r1636/lib/bmx_uci_config/Makefile.orig	2010-04-26 17:49:31.000000000 +0200
+++ bmx-r1636/lib/bmx_uci_config/Makefile	2010-04-26 17:51:23.000000000 +0200
@@ -1,6 +1,14 @@
 
-CFLAGS +=	-fpic -pedantic -Wall -W -Wno-unused-parameter -Os -g3 -std=gnu99 -I../../
-LDFLAGS +=	-shared -luci
+#<agusti> change uci directory, now use variable </agusti>
+UCIPATH = $(shell find ../../../ -maxdepth 1 -type d -name 'uci-*')
+#<sim6> comment next line to add -I../../../uci-0.7.3/ for locate uci.h
+#CFLAGS +=    -fpic -pedantic -Wall -W -Wno-unused-parameter -O1 -g3 -std=gnu99 -I../../
+CFLAGS +=     -fpic -pedantic -Wall -W -Wno-unused-parameter -O1 -g3 -std=gnu99 -I../../ -I$(UCIPATH)
+# </sim6>
+# <sim6> comment next line to add -L../../../uci-0.7.3/ for locate libuci.so
+#LDFLAGS +=   -shared -luci
+LDFLAGS +=    -shared -luci -L$(UCIPATH) 
+# </sim6>
 #-Wl,-soname,bmx_config 
 
 PLUGIN_NAME =   bmx_uci_config
@@ -13,7 +20,10 @@
 PLUGIN_FULLNAME = $(PLUGIN_NAME).so.$(PLUGIN_VER)
 PLUGIN_SHORTNAME = $(PLUGIN_NAME).so
 
-LIBDIR = /usr/lib
+# <sim6> comment next line to add $(INSTALL_PREFIX) for install in kamikaze root directory
+#LIBDIR = /usr/lib
+LIBDIR = $(INSTALL_PREFIX)/usr/lib
+# </sim6>
 
 all:	$(PLUGIN_FULLNAME) Makefile
 
--- bmx-r1636/lib/bmx_gsf_map/Makefile.orig	
+++ bmx-r1636/lib/bmx_gsf_map/Makefile
@@ -13,7 +13,10 @@
 PLUGIN_FULLNAME = $(PLUGIN_NAME).so.$(PLUGIN_VER)
 PLUGIN_SHORTNAME = $(PLUGIN_NAME).so
 
-LIBDIR = /usr/lib
+# <sim6> comment next line to add $(INSTALL_PREFIX) for install in kamikaze root directory
+#LIBDIR = /usr/lib
+LIBDIR = $(INSTALL_PREFIX)/usr/lib
+# </sim6>
 
 all:	$(PLUGIN_FULLNAME) Makefile
 
--- bmx-r1636/lib/bmx_howto_plugin/Makefile.orig	
+++ bmx-r1636/lib/bmx_howto_plugin/Makefile	2009-08-05 17:00:31.000000000 +0200
@@ -13,7 +13,10 @@
 PLUGIN_FULLNAME = $(PLUGIN_NAME).so.$(PLUGIN_VER)
 PLUGIN_SHORTNAME = $(PLUGIN_NAME).so
 
-LIBDIR = /usr/lib
+# <sim6> comment next line to add $(INSTALL_PREFIX) for install in kamikaze root directory
+#LIBDIR = /usr/lib
+LIBDIR = $(INSTALL_PREFIX)/usr/lib
+# </sim6>
 
 all:	$(PLUGIN_FULLNAME) Makefile
 
--- bmx-r1636/lib/bmx_http_info/Makefile.orig	2009-09-30 01:23:41.000000000 +0200
+++ bmx-r1636/lib/bmx_http_info/Makefile	2009-09-30 12:24:53.000000000 +0200
@@ -13,7 +13,10 @@
 PLUGIN_FULLNAME = $(PLUGIN_NAME).so.$(PLUGIN_VER)
 PLUGIN_SHORTNAME = $(PLUGIN_NAME).so
 
-LIBDIR = /usr/lib
+# <sim6> comment next line to add $(INSTALL_PREFIX) for install in kamikaze root directory
+#LIBDIR = /usr/lib
+LIBDIR = $(INSTALL_PREFIX)/usr/lib
+# </sim6>
 
 all:	$(PLUGIN_FULLNAME) Makefile
 
